### PR TITLE
Order loading parent profiles before sub-profiles

### PR DESCRIPTION
**Pull request by ghjs**
Potential fix for [#2259](https://github.com/cobbler/cobbler/issues/2259)

I ran into the same issue and the following are my findings.
```
# cobbler --version
Cobbler 3.1.2
  source: b'97654ed0', b'Tue Oct 13 09:58:06 2020 +0200'
  build time: Thu Oct 15 23:35:55 2020
```
```
# pwd
/var/lib/cobbler/collections/profiles
```
Unsorted directory contents:
```
# ls -1U
CentOS-8.2.2004-x86_64.json.json
minimal-aimpart.json.json
worker-S2600JF.json.json
minimal.json.json
```
minimal has `---distro=CentOS-8.2.2004-x86_64`
minimal-aimpart has `--parent=minimal`
worker-S2600JF has `--parent=minimal-aimpart`
```
# /usr/bin/cobblerd -F
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/cobbler/cobbler_collections/manager.py", line 185, in deserialize
    serializer.deserialize(collection)
  File "/usr/lib/python3.6/site-packages/cobbler/serializer.py", line 125, in deserialize
    storage_module.deserialize(collection, topological)
  File "/usr/lib/python3.6/site-packages/cobbler/modules/serializers/file.py", line 185, in deserialize
    collection.from_list(datastruct)
  File "/usr/lib/python3.6/site-packages/cobbler/cobbler_collections/collection.py", line 211, in from_list
    self.add(item)
  File "/usr/lib/python3.6/site-packages/cobbler/cobbler_collections/collection.py", line 349, in add
    ref.check_if_valid()
  File "/usr/lib/python3.6/site-packages/cobbler/items/profile.py", line 137, in check_if_valid
    raise CX("Error with profile %s - distro is required" % (self.name))
cobbler.cexceptions.CX: 'Error with profile minimal-aimpart - distro is required'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/cobblerd", line 75, in main
    api = cobbler_api.CobblerAPI(is_cobblerd=True)
  File "/usr/lib/python3.6/site-packages/cobbler/api.py", line 106, in __init__
    self.deserialize()
  File "/usr/lib/python3.6/site-packages/cobbler/api.py", line 1526, in deserialize
    return self._collection_mgr.deserialize()
  File "/usr/lib/python3.6/site-packages/cobbler/cobbler_collections/manager.py", line 187, in deserialize
    raise CX("serializer: error loading collection %s: %s. Check /etc/cobbler/modules.conf" % (collection.collection_type(), e))
cobbler.cexceptions.CX: "serializer: error loading collection profile: 'Error with profile minimal-aimpart - distro is required'. Check /etc/cobbler/modules.conf"
```
mv files out/back into /var/lib/cobbler/collections/profiles so the
unsorted file list order will create parent profiles before sub-profiles:
```
# ls -1U
CentOS-8.2.2004-x86_64.json.json
minimal.json.json
minimal-aimpart.json.json
worker-S2600JF.json.json
```
```
# /usr/bin/cobblerd -F
SERVING!
```

This patch loads profiles that have distro specified or sub-profiles where the parent has already been loaded.

### CODE REVIEW

**Comment by SchoolGuy** [comment]
A **big** thank you for this patch!

What I don't understand is how you resolve that a subprofile is loaded after the parent? I mean when the files are in another order, then this is not helping? Or do I miss some magic which takes care of that in your code?

Isn't the only real solution in the end to load everything and then run validations? Or to "fasttrack" the loading of an unresolved profile. Thus switching a bit in the direction of "on-demand" profile and distro loading?

**Comment by ghjs** [comment]
I've looked over the code a bit more and now realize that perhaps the `class Item(object)` `depth` member is intended for this ordered loading from disk?

If so, [try2](https://github.com/ghjs/cobbler/tree/collections_fix_try2) gets me the same result.  If this looks good, I could force update this PR's branch with it.

**Comment by SchoolGuy** [comment]
@ghjs Feel free to experiment but since this is touching the very core of Cobbler I am trying to be a bit more picky as bugs in there can do a lot more arm then in leaf branches of the code...

**Comment by watologo1** [comment]
I have never used sub-profiles and I expect you hit a special case or say: "not that often processed code" and you could be lucky with the ordering as well...
@ghjs : Just curious: Did you experience any other issues with using sub-profiles or does this otherwise work fine?

File: cobbler/cobbler_collections/collection.py [hunk 1]
```diff
@@ -206,9 +206,33 @@ def from_list(self, _list):
...
+            p = profile_list.pop(0)
+            if p.distro != "<<inherit>>" or \
+               self.collection_mgr.profiles().find(name = p.parent) != None:
+                self.add(p)
+                profile_tried.clear()
+            else:
+                try:
+                    profile_tried[p.name] += 1
+                except:
+                    profile_tried[p.name] = 1
```
[the remark this answers is not in the document]
Comment (ghjs) [reply cobbler/cobbler_collections/collection.py:231]: I'm never sure what is more "Pythonic" but went with the EAFP method there...

**Comment by ghjs** [comment]
When I made this branch, I was trying to get my setup going for the exception issue I was seeing.  Now looking closer at the code around my change, I'm more convinced that what I need is already there, but is either commented out or perhaps wrong for python3 list.sort().  I've been running [this](https://github.com/ghjs/cobbler/commit/4c5f340d2e004cedf62a3cfd36fe096e8869245a) which I feel is the correct approach now.  The commit log has all the details.  I was tempted to force update this PR branch with it, but will delay so that commit can be looked over again.

**Comment by SchoolGuy** [comment]
@ghjs So how do you want to proceed? I have to honestly say that I don't have the resources time-related to set such things up. At least not this week. I am open for suggestions and would really love if we could write a test which proves that this is now not working and after your commit working! Thus that we can be sure that this problem never bothers anyone again...

**Comment by ghjs** [comment]
@SchoolGuy I've force pushed my try2 branch over my initial PR branch.

I feel it is the correct approach that restores the original intention of the collection serializer code.  Ie, the '#FIXME' is addressed from 894be5e1d6cacd56891ec0a9c55bca431f03684e where other sort issues were addressed but not the one at hand.

As the code sits now, the loading of Item objects is completely arbitrary, which is the problem I guess I (and others) ran into randomly.  For the file serializer, the loading order is dependent on the used filesystem's internal dirent structure and implementation of readdir() followed by python's glob.glob() before the list is returned.  However mongodb returns data, I'm not sure; the existing code reflects the same topological ordering logic should be applied however.

The overall (restorative) change is pretty minimal, it's basically taking the arbitrary ordered list of dicts that was loaded from the configured serializer and sorting them by the dict's depth value before iterating the sorted list for object initialization.

As for a test case to prove failing code to working code with this commit, that's a difficult thing to prove or otherwise enforce on an OS environment the test suite has little control over (i.e. how the FS returns data during readdir()).  Again, the source data is loaded in arbitrary order by the OS environment and the cobbler code to enforce the necessary loading order of objects has been disabled (commented out), or if the code walked the mongodb serializer path, would cause an exception for python3's list.sort() method has no cmp argument.

I've added (restored really) Docstring comments to the topological parameter with the latest commit, as I feel the context or reasons for it have been lost with code refactoring along with, perhaps, the migration of cobbler to python3.

@watologo1 I'm coming from cobbler 2.6 and 2.8 where I've used subprofiles quite a bit.  Basically we setup a profile to point at the major.minor OS distro perhaps with data center specific profile attributes.  Then we have subprofiles that set ks-meta (autoinstall-meta now) along with other machine specific profile attributes.  When the time comes to do some distro level upgrade, only the parent profile needs to be updated and all the various machine types in the data center will point to the next distro release when we get around to re-kickstarting them.  So, I'm very interested in keeping this functionality working in cobbler, if possible.

To both of you: Thank you for all of your efforts in keeping cobbler alive/active!

**Review by SchoolGuy — APPROVED** [review]
Since a testcase is nearly impossible to write and we have docstrings explaining this, I will give my approval and thanks for this!

**Review by brejoc — APPROVED** [review]
lgtm

**Comment by SchoolGuy** [comment]
We got two approvals. Merged.

**Pull request merged**